### ISSUE TITLE
consider updating the base64 third-party to libb64 2.0.0 or to cppcodec

### ISSUE THREAD

**Issue by GaryElshaw** [transmission/transmission#2997]
Nine years old. @ckerr The last released version 2.0.0 was 27 February, 2020 and compiles fine on MacOS Monterey using the same cmakefile Transmission currently uses. The last commit to the repository in September 2020 was from some guy called @mikedld. The current repository, as at 10 September 2020 also compiles fine with the old cmakefile in the libb64 directory.

I tried to add the directory to my fork and do a PR, but no dice. Thankfully the old libb64 directory can be completely overwritten while retaining the old base cmakefile.

gh repo clone libb64/libb64

**Comment by mikedld** [comment]
What exactly is the issue here? Something like base64 algorithm doesn't really age, once written and tested it can be used for decades...

**Comment by GScottElshaw** [comment]
Almost everything in 'third-party' has been neglected, in some cases, for between _6 - 12 years_ and i think that is an issue of, at the least, bad housekeeping.

**Comment by ckerr** [comment]
@GScottElshaw agreed that it's worth looking at each of the dependencies to see if there's a case for upgrading them, but  please keep it on a per-dependency basis. Saying "Everything has been neglected for 6-12 years" is not an argument for updating libb64. Moreover, saying that this is an issue of bad housekeeping is inaccurate:

looking at https://github.com/libb64/libb64/blob/master/CHANGELOG.md

- @mikedld applied the 1.3.0 bugfix in https://github.com/transmission/libb64/commit/8456a5b5c99ef4544a2ed8e8f20a2fb673dd4192
- @mikedld applied the 1.4.0 overflow bugfix in https://github.com/transmission/libb64/commit/c30e15786beea42a84ae15f3dd8dff9bbd934547

I don't see any 2.0.0 features that we would use So what is the benefit in updating to 2.0.0? I see two effects:

- Would let us drop two of our patches, making us slightly closer to upstream. We'd only be adding CMakeLists.txt
- Fewer warnings during compiling in strict mode, e.g. on MSVC

Actually in the case of libb64, I'd be more interested in replacing it with https://github.com/tplgy/cppcodec since libb64 has a (minor) extra cost of us maintaining cmake files for it. Since cppcodec is header-only and uses cmake anyway, that cost would go away.

**Comment by GScottElshaw** [comment]
I don't see a downside, but if it also prompts a discussion about alternatives that may be better for Transmission, that's all great.

I feel a bit sad to be honest.

**Comment by ckerr** [comment]
> there's e.g. no way to skip building examples and tools, and installing their files

I meant that it's a header-only library, so we would neither need to add new makefiles nor use their makefiles

**Comment by ckerr** [comment]
> I feel a bit sad to be honest.

This isn't a value judgement on the dependencies, or on these issues, or on you, so don't feel too bad. :green_heart: 

The thing to consider that on most Linux systems we try and pull in the system version of the library instead of a bundled version. By keeping the dependency's minimum version low, we help Transmission's binary to be smaller on older Linux distributions because even on those old systems we can use the system library instead of a custom-built one.

So basically the rule of thumb is: does the new version of the dependency have performance / memory improvements we'll benefit from, or new features we can  use, or fix a bug that affects us? If so, then we should consider updating. If the changes don't affect us, then we're better off leaving the dependency's minimum version number low.

**Comment by Coeur** [comment]
Done in #3960. So I think we can close this issue, right?
